keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,47 +173,52 @@ to set the feature gate to true to gain these this Secure by Default benefit.
 
 ### Risks and Mitigations
 
-Image authentications with a registry may expire. To mitigate expirations a
-a timeout will be used to force re-authentication. The timeout could be a
-container runtime feature or a `kubelet` feature. If at the container runtime,
-images would not be present during the EnsureImagesExist step, thus would have
-to be pulled and authenticated if necessary. This timeout feature will be
-implemented in alpha.
-
-Since images can be pre-loaded, loaded outside the `kubelet` process, and
-garbage collected.. the list of images that required authentication in `kubelet`
-will not be a source of truth for how all images were pulled that are in the
-container runtime cache. To mitigate, images can be garbage collected at boot.
-And for alpha, we will not persist ensure metadata across reboot of host, and restart
-of kubelet, and possibly look at a way to add ensure metadata for images loaded
-outside of kubelet. In beta we will add a switch to enable re-auth on boot for
-admins seeking that instead of having to garbage collect where they do not use
-or expect preloaded images since boot.
+- Image authentications with a registry may expire.
+  - To mitigate expirations a timeout will be used to force re-authentication.
+    This timeout will be configured as a kubelet configuration field `pullImageSecretRecheckPeriod`.
+    This timeout feature will be implemented in alpha.
+
+- Images can be "pre-loaded", or pulled behind the kubelet's back before it starts.
+  In this case, the kubelet is not managing the credentials for these images.
+  - To mitigate, metadata will be persisted across reboot. The kubelet will compare previously
+    cached credentials against the images that exist. On a new image pull, the kubelet will use
+    its saved cache and revalidate as necessary.
+    In other words: even if the images are already cached, if new images are present that have not
+    previously been authenticated against a pods credentials, then the image will be revalidated.
```

Comment:
Before beta, there should be further integration with the container runtimes to provide a validation/credentials process to support authentication/re-authentication of the pre-loaded images.

Reply:
do you mean a new CRI RPC other than PullImage?

Reply:
> we could do something different just for pre-loads and never

what's the difference between an auth check and a pull if the blob is already present on disk? We should do practically the same. 

> The ensure process being split as it is between kubelet, the container runtime, and the registries.. just doesn't make a lot of sense

fair, but I don't see this KEP as being the space to redefine this. I would suggest we scope this to fixing the existing cases and consider new integrations as a separate KEP

Reply:
agreed.